api-server/ui/Map.tsx

```diff
@@ -1,27 +1,46 @@
+import { SKContext } from '@chacal/signalk-ts'
 import * as React from 'karet'
 import * as U from 'karet.util'
+import { Observable } from 'kefir'
+import { Atom } from 'kefir.atom'
 import { LatLngBounds, LeafletEvent } from 'leaflet'
+import * as L from 'partial.lenses'
 import { GeoJSON, Map as LeafletMap, TileLayer } from 'react-leaflet'
 import { Coords, TrackGeoJSON } from '../domain/Geo'
-import { Atomized } from './ui-domain'
+import updateTracksFor from './trackprovider'
+import { LoadState, mapArrayInObs, Vessel } from './ui-domain'
 
 // Lift LeafletMap to Karet to support Atoms as props
 const KaretMap = U.toKaret(LeafletMap)
 
+export interface Track {
+  context: SKContext
+  geoJson: TrackGeoJSON
+}
+
 export interface MapProps {
-  center: Coords
-  zoom: number
-  bounds: LatLngBounds
-  tracks: TrackGeoJSON
+  center: Observable<Coords, any>
+  zoom: Atom<number>
+  bounds: Atom<LatLngBounds>
+  vessels: Atom<Vessel[]>
 }
 
-const Map = ({ center, zoom, bounds, tracks }: Atomized<MapProps>) => {
+const Map = ({ center, zoom, bounds, vessels }: MapProps) => {
   const updateBoundsFromMap = (comp: LeafletMap) =>
     bounds.set(comp.leafletElement.getBounds())
   const updateBoundsFromEvent = (e: LeafletEvent) => {
     bounds.set(e.target.getBounds())
     zoom.set(e.target.getZoom())
   }
+  const vesselsWithTracks = vessels.map(vessels =>
+    vessels.filter(vesselHasTracks)
+  )
+
+  // Update tracks in global state when vessels change
+  updateTracksFor(vessels, zoom, bounds)
+
+  // Reload tracks when bounds or zoom changes
+  bounds.merge(zoom).onValue(() => forceTrackReload(vessels))
```

Comment:
You should avoid `onValue` in Calmm applications.  The main problem is that the subscription may be leaked.  In this case the subscription will be live as long as either `bounds` or `zoom` is alive.  Inside a component you can avoid calling `onValue` by embedding side-effecting observables into the VDOM.

Reply:
So, what you could do here is to embed something like the following

```
{U.consume(() => forceTrackReload(vessels), bounds.merge(zoom))}
```

to the VDOM `return`ed below.  [`U.consume`](https://github.com/calmm-js/karet.util#u-consume) returns an observable and this way Karet will automatically unsubscribe from the observable when the component is unmounted.  There are also other helpers like [`U.set`](https://github.com/calmm-js/karet.util#u-set) that can be used for side-effects embedded to VDOM.

Reply:
Thx! Fixed. Was completely unaware of the problems with the leaking subscriptions.. :/